File: src/void/vweb/engine/platform.ts
export type Platform = {
  // linear memory shared between Go and JS.
  memory: WebAssembly.Memory
  // initialises the Go runtime and calls `main()`.
  _start(): void
  // byte offset into `memory` of the poll.
  PollPtr(): number
  BeepPtr(): number
  BeepCount(): number
  // consumes a pending fullscreen req.
  FullscreenReq(): number
  ScreenshotReq(): number
  ContextLossReq(): number
  DrawAlways(): number
  DrawOnBlur(): number
  ReqWakelock(): number
  RenderMode(): number
  UpdateInMillisReq(): bigint
  LayerConfigsPtr(): number
  Update(): Loop
  // byte offset into `memory` of the first tile.
  BoardTilesPtr(): number
  // board size in pixels. origin is always `(0, 0)`.
  BoardW(): number
  BoardH(): number
  // tile pixel dimensions.
  BoardTileW(): number
  BoardTileH(): number
  // camera position in world pixels.
  CamX(): number
  CamY(): number
  AtlasAnimCount(): number
  AtlasCelsPerAnim(): number
  AtlasCelsPtr(): number
  AtlasCelsCount(): number
}

export type Loop = typeof LoopPause | typeof LoopLoop
export const LoopPause = 0 as const
export const LoopLoop = 1 as const

export const RenderModeFloat = 0 as const
export const RenderModePixel = 1 as const
export const FullscreenReqNone = 0 as const
export const FullscreenReqExit = 1 as const
export const FullscreenReqEnter = 2 as const
export const FullscreenReqPortrait = 3 as const
export const FullscreenReqLandscape = 4 as const

File: src/void/vengine/platform.go
package vengine

type Platform interface {
	PollPtr() uintptr
	BeepPtr() uintptr
	BeepCount() uint32
	FullscreenReq() int32
	ScreenshotReq() int32
	ContextLossReq() int32
	DrawAlways() int32
	DrawOnBlur() int32
	RenderMode() int32
	UpdateInMillisReq() uint64
	LayerConfigsPtr() uintptr
	Update() Status
	BoardTilesPtr() uintptr
	BoardW() int32
	BoardH() int32
	BoardTileW() uint8
	BoardTileH() uint8
	CamX() float32 // to-do: rename Phy.
	CamY() float32
	AtlasAnimCount() uint32
	AtlasCelsPerAnim() uint32
	AtlasCelsPtr() uintptr
	AtlasCelsCount() uint32
}
